File: 1160-letter-tile-possibilities/1160-letter-tile-possibilities.cpp
class Solution {
public:
    int numTilePossibilities(string tiles) {
        unordered_set<string> seq;
        vector<bool> used(tiles.length(), false);

        generateSeq(tiles, "", used, seq);

        return seq.size() - 1;
    }

private:
    void generateSeq(string& tiles, string current, vector<bool>& used,
                     unordered_set<string>& seq) {
        seq.insert(current);
        for (int pos = 0; pos < tiles.length(); ++pos) {
            if (!used[pos]) {
                used[pos] = true;
                generateSeq(tiles, current + tiles[pos], used, seq);
                used[pos] = false;
            }
        }
    }
};

File: 1160-letter-tile-possibilities/1160-letter-tile-possibilities.rs
use std::collections::HashMap;

impl Solution {
    pub fn num_tile_possibilities(tiles: String) -> i32 {
        let mut char_cnt = HashMap::new();
        for ch in tiles.chars() {
            *char_cnt.entry(ch).or_insert(0) += 1;
        }
        Self::find_seq(&mut char_cnt)
    }
    pub fn find_seq(char_cnt: &mut HashMap<char, i32>) -> i32 {
        let mut cnt = 0;

        for (&ch, &freq) in char_cnt.clone().iter() {
            if freq == 0 {
                continue;
            }

            cnt += 1;
            *char_cnt.get_mut(&ch).unwrap() -= 1;
            cnt += Self::find_seq(char_cnt);
            *char_cnt.get_mut(&ch).unwrap() += 1;
        }

        cnt
    }
}